### PR TITLE
Issue #SB-4561: Notification in review workflow

### PR DESCRIPTION
**Pull request by souravdey091**

### CODE REVIEW

File: src/middlewares/request.middleware.js [hunk 1]
```diff
@@ -12,6 +12,7 @@ var reqMsg = messageUtil.REQUEST
 var responseCode = messageUtil.RESPONSE_CODE
 var apiVersions = messageUtil.API_VERSION
 var filename = path.basename(__filename)
+const jwt = require('jsonwebtoken')
```
Comment (anoophm) [inline src/middlewares/request.middleware.js]: use var

File: src/service/emailService.js [hunk 2]
```diff
@@ -209,62 +212,210 @@ function rejectFlagContentEmail (req, callback) {
...
+      if (err || result.responseCode !== responseCode.SUCCESS) {
+        callback(new Error('Form API failed'), null)
+      } else {
+        callback(null, result)
+      }
+    })
+  }
+}
+
+function constructLiveUrl (content) {
```
Comment (anoophm) [inline src/service/emailService.js]: getPublisedContentUrl

File: src/service/emailService.js [hunk 3]
```diff
@@ -209,62 +212,210 @@ function rejectFlagContentEmail (req, callback) {
...
+  var baseUrl = configUtil.getConfig('CONTENT_REPO_BASE_URL')
+  if (content.mimeType === 'application/vnd.ekstep.content-collection') {
+    if (content.contentType !== 'Course') {
+      return `${baseUrl}/resources/play/collection/${content.identifier}`
```
Comment (anoophm) [inline src/service/emailService.js]: dont use `${}`

File: src/service/emailService.js [hunk 4]
```diff
@@ -209,62 +212,210 @@ function rejectFlagContentEmail (req, callback) {
...
+      return `${baseUrl}/learn/course/${content.identifier}`
+    }
+  } else if (content.mimeType === 'application/vnd.ekstep.ecml-archive') {
+    return `${baseUrl}/resources/play/content/${content.identifier}`
+  } else {
+    return `${baseUrl}/resources/play/content/${content.identifier}`
+  }
+}
+
+function constructDraftUrl (content) {
```
Comment (anoophm) [inline src/service/emailService.js]: getDraftContentUrl

File: src/service/emailService.js [hunk 5]
```diff
@@ -209,62 +212,210 @@ function rejectFlagContentEmail (req, callback) {
...
+  var genricMimeType = [
+    'application/pdf', 'video/mp4', 'video/x-youtube', 'video/youtube',
+    'application/vnd.ekstep.html-archive', 'application/epub',
+    'application/vnd.ekstep.h5p-archive', 'video/webm', 'text/x-url'
+  ]
+  var baseUrl = `${configUtil.getConfig('CONTENT_REPO_BASE_URL')}/workspace/content/edit`
```
Comment (anoophm) [inline src/service/emailService.js]: dont use `${}`

File: src/service/emailService.js [hunk 6]
```diff
@@ -209,62 +212,210 @@ function rejectFlagContentEmail (req, callback) {
...
+    }
+  } else if (content.mimeType === 'application/vnd.ekstep.ecml-archive') {
+    return `${baseUrl}/resources/play/content/${content.identifier}`
+  } else {
+    return `${baseUrl}/resources/play/content/${content.identifier}`
+  }
+}
+
+function constructDraftUrl (content) {
+  var genricMimeType = [
```
Comment (anoophm) [inline src/service/emailService.js]: remove this

File: src/service/emailService.js [hunk 7]
```diff
@@ -209,62 +212,210 @@ function rejectFlagContentEmail (req, callback) {
...
+    'application/vnd.ekstep.html-archive', 'application/epub',
+    'application/vnd.ekstep.h5p-archive', 'video/webm', 'text/x-url'
+  ]
+  var baseUrl = `${configUtil.getConfig('CONTENT_REPO_BASE_URL')}/workspace/content/edit`
+  if (content.mimeType === 'application/vnd.ekstep.content-collection') {
+    return `${baseUrl}/collection/${content.identifier}/${content.contentType}/draft/
+    ${content.framework}`
+  } else if (content.mimeType === 'application/vnd.ekstep.ecml-archive') {
+    return `${baseUrl}/content/${content.identifier}/draft/${content.framework}`
+  } else if ((genricMimeType).includes(content.mimeType)) {
```
Comment (anoophm) [inline src/service/emailService.js]: just else

File: src/service/emailService.js [hunk 8]
```diff
@@ -209,62 +212,210 @@ function rejectFlagContentEmail (req, callback) {
...
  * @param {function} callback
  */
 function publishedContentEmail (req, callback) {
-  var data = req.body
-  data.contentId = req.params.contentId
-  var rspObj = req.rspObj
-
-  if (data.contentId) {
+  if (!req.params.contentId) {
     callback(new Error('Content id is missing'), null)
```
Comment (anoophm) [inline src/service/emailService.js]: add log

File: src/service/emailService.js [hunk 9]
```diff
@@ -209,62 +212,210 @@ function rejectFlagContentEmail (req, callback) {
...
   }
+  var rspObj = req.rspObj
```
Comment (anoophm) [inline src/service/emailService.js]: not needed

File: src/service/emailService.js [hunk 10]
```diff
@@ -209,62 +212,210 @@ function rejectFlagContentEmail (req, callback) {
...
+        callback(new Error('Form API failed'), null)
+      } else {
+        callback(null, result)
+      }
+    })
+  }
+}
+
+function constructLiveUrl (content) {
+  var baseUrl = configUtil.getConfig('CONTENT_REPO_BASE_URL')
```
Comment (anoophm) [inline src/service/emailService.js]: take base url from req, like in unlisted

File: src/service/emailService.js [hunk 11]
```diff
@@ -209,62 +212,210 @@ function rejectFlagContentEmail (req, callback) {
...
+  if (!req.params.contentId) {
     callback(new Error('Content id is missing'), null)
   }
+  var rspObj = req.rspObj
+  var formRequest = {
+    request: {
+      'type': 'notification',
+      'action': 'publish',
+      'subType': 'email',
+      'rootOrgId': req.headers['x-channel-id']
```
Comment (anoophm) [inline src/service/emailService.js]: use req.get

File: src/service/emailService.js [hunk 12]
```diff
@@ -209,62 +212,210 @@ function rejectFlagContentEmail (req, callback) {
...
+      body = body.replace(/{{Content type}}/g, cData.contentType)
+        .replace(/{{Content title}}/g, cData.name)
+        .replace(/{{Creator name}}/g, req.headers['userName'])
+      var lsEmailData = {
+        request: getEmailData(null, subject, body, null, null, null,
+          null, data.templateConfig.result.form.data.templateName, eData.logo)
+      }
+      lsEmailData.request.recipientSearchQuery = {
+        'filters': {
+          'channel': req.headers['x-channel-id'],
```
Comment (anoophm) [inline src/service/emailService.js]: req.get

File: src/service/emailService.js [hunk 13]
```diff
@@ -19,7 +20,8 @@ var responseCode = messageUtils.RESPONSE_CODE
  * @param {string} recipientUserIds
  * @param {string} emailTemplateType
```
Comment (bvinayakumar) [inline src/service/emailService.js]: imageUrl is missed in comment

File: src/service/emailService.js [hunk 14]
```diff
@@ -209,62 +213,229 @@ function rejectFlagContentEmail (req, callback) {
...
-    function (res) {
-      LOG.info(utilsService.getLoggerData(rspObj, 'INFO', filename, 'rejectFlagContentEmail',
-        'Email sent successfully', rspObj))
+/**
+ * Below function is used for send email when review content api called
+ * @param {object} req
+ * @param {function} callback
+ */
+function reviewContentEmail (req, callback) {
+  if (!req.params.contentId) {
```
Comment (bvinayakumar) [inline src/service/emailService.js]: Duplicated code block (publishedContentEmail)

File: src/service/emailService.js [hunk 15]
```diff
@@ -209,62 +213,229 @@ function rejectFlagContentEmail (req, callback) {
...
+    request: {
+      'type': 'notification',
+      'action': 'sendForReview',
+      'subType': 'email',
+      'rootOrgId': req.get('x-channel-id')
+    }
+  }
+
+  async.waterfall([
+    function (callback) {
```
Comment (bvinayakumar) [inline src/service/emailService.js]: Duplicated code block (publishedContentEmail)

File: src/service/emailService.js [hunk 16]
```diff
@@ -209,62 +213,229 @@ function rejectFlagContentEmail (req, callback) {
...
+        request: getEmailData(null, subject, body, null, null, null,
+          null, data.templateConfig.result.form.data.templateName, eData.logo)
+      }
+      lsEmailData.request.recipientSearchQuery = {
+        'filters': {
+          'channel': req.get('x-channel-id'),
+          'organisations.roles': ['CONTENT_REVIEWER']
+        }
+      }
+      contentProvider.sendEmail(lsEmailData, req.headers, function (err, res) {
```
Comment (bvinayakumar) [inline src/service/emailService.js]: Duplicated code block (publishedContentEmail)

File: src/service/emailService.js [hunk 17]
```diff
@@ -331,9 +511,9 @@ function rejectContentEmail (req, callback) {
  * @param  {[String]} identifier [contentID]
  * @return {[String]}         [base64 string]
  */
-var getBase64Url = function (type, identifier) {
-  return Buffer.from(type + '/' + identifier).toString('base64')
-}
+// var getBase64Url = function (type, identifier) {
```
Comment (bvinayakumar) [inline src/service/emailService.js]: Why is this code commented as part of this PR?

File: src/service/emailService.js [hunk 18]
```diff
@@ -273,56 +444,65 @@ function publishedContentEmail (req, callback) {
  * @param {function} callback
  */
 function rejectContentEmail (req, callback) {
-  var data = req.body
-  data.contentId = req.params.contentId
-  var rspObj = req.rspObj
-
-  if (data.contentId) {
+  if (!req.params.contentId) {
```
Comment (bvinayakumar) [inline src/service/emailService.js]: Avoid duplicated code in this method too.

File: src/service/emailService.js [hunk 19]
```diff
@@ -331,26 +422,13 @@ function rejectContentEmail (req, callback) {
  * @param  {[String]} identifier [contentID]
  * @return {[String]}         [base64 string]
  */
```
Comment (bvinayakumar) [inline src/service/emailService.js]: Remove getBase64Url comment as well.

**Pull request merged**